### PR TITLE
User Interface: deployment using dockerized static site

### PR DESCRIPTION
**Pull request by eap**
* Add a pre-loading config system for the UI that works
   when being served from a static server (like nginx).

 * Add a build system for the UI that creates an env
   controlled container.

 * Allow additional scopes to be added during auth.

 * Update deploy script to show how the new front-end
   container is used.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Comment by eap** [comment]
jenkins retest

**Comment by eap** [comment]
This is ready for review.

File: ui/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# Create a container that will configure the front-end and serve from a static directory.
```
Comment (calbach) [inline ui/Dockerfile]: Is this for local development, or do you mean to deploy this?

File: ui/public/index.html [hunk 2]
```diff
@@ -8,6 +8,7 @@
       manifest.json provides metadata used when your web app is added to the
       homescreen on Android. See https://developers.google.com/web/fundamentals/engage-and-retain/web-app-manifest/
     -->
+    <script src="loadConfig.js"></script>
```
Comment (calbach) [inline ui/public/index.html]: nit: move above the manifest comment

File: ui/public/loadConfig.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+// root of the server and loads it into the global `GlobalReactConfig`.
+(function () {
+    var xobj = new XMLHttpRequest();
+    xobj.overrideMimeType("application/json");
+    // Run this in synchronous mode since the configuration should be loaded
+    // before the React app initializes.
+    xobj.open('GET', 'config.json', false);
+    xobj.onreadystatechange = function () {
+        if (xobj.readyState == 4 && xobj.status == "200") {
+            window.GlobalReactConfig = JSON.parse(xobj.responseText);
```
Comment (calbach) [inline ui/public/loadConfig.js]: Shouldn't there be some kind of callback triggered here to block bootstrapping of the app? What guarantee is there that config.json will be loaded by the time your react code runs?

File: ui/public/config.json [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  "FILTER_TO_CURRENT_USER_COMMENT": "boolean, whether to filter the cluster list to the logged-in user.",
+
+  "DEPLOY_ENVIRONMENT": "local",
+  "DEPLOY_ENVIRONMENT_COMMENT": "Uses mock APIs when set to 'local': allowed values 'local' and 'prod'.",
+
+  "OAUTH_CLIENT_ID": "---client-id---",
+  "OAUTH_CLIENT_ID_COMMENT": "OAuth client ID",
+
+  "DEFAULT_PROJECT": "---project-id---",
+  "DEFAULT_PROJECT_COMMENT": "Provide a default project for users (populated in cluster creation form).",
```
Comment (calbach) [inline ui/public/config.json]: Are any/all of these optional? Worth specifiying

File: ui/public/loadConfig.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+// This function accesses a file "config.json" expected to be at the
+// root of the server and loads it into the global `GlobalReactConfig`.
+(function () {
+    var xobj = new XMLHttpRequest();
+    xobj.overrideMimeType("application/json");
+    // Run this in synchronous mode since the configuration should be loaded
+    // before the React app initializes.
+    xobj.open('GET', 'config.json', false);
+    xobj.onreadystatechange = function () {
+        if (xobj.readyState == 4 && xobj.status == "200") {
```
Comment (calbach) [inline ui/public/loadConfig.js]: hmm.. won't this be called with the other readystates before this? (thereby triggering your console.log statement?). In the JMUI we added callbacks on `onload()` and `onerror()` here (I haven't dug too deploy as to whether that's 100% the right thing to do).

File: automation/gce-deploy/site.conf [hunk 6]
```diff
@@ -53,26 +43,30 @@ LDAPCacheTTL ${LDAP_CACHE_TTL}
...
-        Header unset Access-Control-Allow-Methods "expr=%{REQUEST_URI} !~ m#^/notebooks#"
-        Header always set Access-Control-Allow-Methods "GET,POST,PUT,PATCH,DELETE,OPTIONS,HEAD" "expr=%{REQUEST_URI} !~ m#^/notebooks#"
-        Header unset Access-Control-Max-Age "expr=%{REQUEST_URI} !~ m#^/notebooks#"
-        Header always set Access-Control-Max-Age 1728000 "expr=%{REQUEST_URI} !~ m#^/notebooks#"
-
-        RewriteEngine On
-        RewriteCond %{REQUEST_METHOD} OPTIONS
-        RewriteCond %{REQUEST_URI} !^/notebooks
-        RewriteRule ^(.*)$ $1 [R=204,L]
+    <Location "/">
```
Comment (calbach) [inline automation/gce-deploy/site.conf]: I haven't looked at this in detail, I'll leave that to someone who is already familiar with what this config was being used for.

File: automation/gce-deploy/docker-compose.yml [hunk 7]
```diff
@@ -28,6 +28,17 @@ services:
     links:
       - sqlproxy:sqlproxy
 
+  userinterface:
+      image: gcr.io/TEMPLATE_VAR_DOCKER_PROJECT/leonardo-ui:TEMPLATE_VAR_DOCKER_TAG
```
Comment (calbach) [inline automation/gce-deploy/docker-compose.yml]: Will this be substituted? I thought you need some kind of `$` syntax for docker-compose env substitutino.

File: ui/public/loadConfig.js (hunk 5 above)
Comment (eap) [reply ui/public/loadConfig.js]: good catch. I solved this a different way that now ensures that the config is loaded before loading config dependent parts of the site. See src/components/ConfigLoader.js

File: ui/public/index.html (hunk 2 above)
Comment (eap) [reply ui/public/index.html]: removed entirely and put into main react app.

File: ui/src/components/ConfigLoader.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+  constructor(props) {
+    super(props);
+    this.state = {configLoading: true};
+  }
+
+  /**
+   * Load configuration from static assets.
+   */
+  loadConfig = () => {
```
Comment (calbach) [inline ui/src/components/ConfigLoader.js]: nit: `loadConfig() {` ? (to match render() below)

File: ui/build.sh [hunk 9]
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env bash
+
+# Note: in most cases this script should be invoked by
+# the build script in "leonardo/docker/build.sh".
```
Comment (calbach) [inline ui/build.sh]: I don't see that script modified in this PR, should it have been?

**Pull request merged**